engine/src/flutter/lib/web_ui/lib/src/engine/web_paragraph/painter.dart

```diff
@@ -38,6 +38,27 @@ abstract class Painter {
 
   /// Paints the decorations previously filled by [fillDecorations].
   void paintDecorations(ui.Canvas canvas, ui.Rect sourceRect, ui.Rect targetRect);
+
+  /// Adjust the _paintCanvas scale based on device pixel ratio
+  double adjustCanvas(ui.Canvas canvas, double width, double height) {
+    if (width == 0 || height == 0) {
+      return 1.0;
+    }
+
+    final double zoomFactor = ui.window.devicePixelRatio;
+    _paintCanvas.width = (width * zoomFactor).ceilToDouble();
+    _paintCanvas.height = (height * zoomFactor).ceilToDouble();
+    paintContext.scale(zoomFactor, zoomFactor);
+
+    WebParagraphDebug.log(
+      'adjustPaintCanvas: ${_paintCanvas.width}x${_paintCanvas.height} @ $zoomFactor',
+    );
+    return zoomFactor;
+  }
+
+  void clearCanvas(ui.Rect sourceRect) {
+    paintContext.clearRect(0, 0, _paintCanvas.width!.ceil(), _paintCanvas.height!.ceil());
+  }
 }
```

Comment:
Suggestions based on offline discussion:

```suggestion
/* Globals */
const _kPaintWidth = 100;
const _kPaintHeight = 100;

final DomOffscreenCanvas _paintCanvas = createDomOffscreenCanvas(_kPaintWidth, _kPaintHeight);
final _paintContext = (_paintCanvas.getContext('2d')! as DomCanvasRenderingContext2D)..save();
double? _currentDpr;

abstract class Painter {
  // ....
  void resizePaintCanvas(ui.Canvas canvas, double dpr) {
    if (dpr == _currentDpr) {
      return;
    }

    _paintCanvas.width = (_kPaintWidth * dpr).ceilToDouble();
    _paintCanvas.height = (_kPaintHeight * dpr).ceilToDouble();
    
    if (_currentDpr != null) {
      paintContext.restore();
    }
    paintContext.scale(dpr, dpr);
    paintContext.save();
    
    _currentDpr = dpr;
  }
}
```